cpp/src/arrow/python/arrow_to_python.cc

```diff
@@ -112,9 +112,17 @@ Status GetValue(PyObject* context, const UnionArray& parent, const Array& arr,
     case Type::BOOL:
       *result = PyBool_FromLong(checked_cast<const BooleanArray&>(arr).Value(index));
       return Status::OK();
-    case Type::INT64:
+    case Type::INT64: {
+#if PY_MAJOR_VERSION < 3
+      const std::string& child_name = parent.type()->child(type)->name();
+      if (child_name == "py2_int") {
+        *result = PyInt_FromSsize_t(checked_cast<const Int64Array&>(arr).Value(index));
+        return Status::OK();
+      }
+#endif
```

Comment:
I think it's slightly clearer to do

```c++
#if PY_MAJOR_VERSION < 3
  // handle Py2 case
#else
  // handle Py3 case
#endif
```

instead of

```c++
#if PY_MAJOR_VERSION < 3
  // handle Py2 case
#endif
// handle Py3 case
```

Reply:
There is only a Python 2 case here, the "Python 3" case needs to be handled in both Python 2 and Python 3 (in Python 2 there is both int and long).

Reply:
I see, actually, now that I think about it, we don't need the `#if`, right?

Reply:
It's needed to get it to compile, since in Python 3 the function PyInt_FromSsize_t doesn't exist. Arrow has some common code to handle differences like this but I didn't find anything that covers this function.

Reply:
Ok, sounds good.